docs/design/2024-05-10-extension-authentication-plugin.md

```diff
@@ -0,0 +1,231 @@
+
+### Functional Tests
+
+We will implement unit tests for all of the auth plugin extension interfaces, and ensure that they are executed properly in the code path that we inject them.
+
+### Scenario Tests
+
+We will implement client to server tests to test out scenarios where MySQL users are created with and without using an auth plugin from the extension system. For example:
++ Creating a user using the plugin
+  + Connecting as the user and verify that the connection verification API is called and errors are correctly translated to connection rejections
+    + Execute SQL statements and verify that the request (dynamic) verification APIs are called and errors are correctly translated to permission rejections
+    + Alter user changes can be applied correctly
+    + Passwords can be encoded with the API defined in the plugin
++ When the connection changes between users, the system should behave correctly
+  + From non-plugin user to plugin user
+  + From plugin user to non-plugin user
+  + From plugin user to plugin user
+
+### Compatibility Tests
+
+This feature should integrate well with all the existing test cases since it is an additional functionality. Existing tests related to authentication and authorization should continue to pass since they are not using auth plugins.
+
+### Benchmark Tests
+
+This feature should not affect the performance of the TiDB as it simply provides interfaces for users to inject their own auth plugins via the extension system. The performance implication will come from the implementation of those plugins which depends on how the users of the framework want to use it.
+
+## Impacts & Risks
+
+If a user is using an auth plugin, the SQL privilege status in TiDB might not reflect the actual privileges, since there might be an outside system that decides whether the privilege check can actually pass. However, this is an intentional behavioral change when the user uses an auth plugin since we are providing users customizations on checks to run.
+
+The design does not account for implementing the same level of proxy user support as in MySQL auth plugin, since currently TiDB does not support [proxy users](https://dev.mysql.com/doc/extending-mysql/8.0/en/writing-authentication-plugins-proxy-users.html) yet. Also existing support for password expiry and validation will not work, as mentioned in the earlier section.
+
+
+## Investigation & Alternatives
+
+### MySQL
+
+MySQL supports customizing authentication methods with [authentication plugins](https://dev.mysql.com/doc/extending-mysql/8.0/en/writing-authentication-plugins-server-side.html). This design proposes a similar set of interfaces for users to define, compared to those in MySQL. The main differences are that we also allow defining authorization checks (SQL privileges).
```

Comment:
I am not pretty sure the necessity of customizing authorization checks, especially `VerifyPrivilege`. Normally for a basic `SELECT` statement, only `SELECT` privilege is required. Checking other privileges, like `INSERT`, for a read-only `SELECT` statement would be weird. Since `mysql.PrivilegeType` is not extensible, is there any scenario that additional privileges would be required by customizing authenrization?

Reply:
There could be use cases where there is a centralized privilege management system which the user would like to integrate TiDB with. In this case, the user could make RPC calls to that system and make a final authorization decision based on that (since based on the design, at this point user already has enough privilege in MySQL).

Basically users can implement their own customized authorization logic on top of the MySQL privilege checks. If the statement requires only `SELECT` privilege, then `VerifyPrivilege` will only receive one call with the `SELECT` privilege according to the current implementation in TiDB, but it's true that there's nothing stopping the user to check for `INSERT` privilege if they somehow actually want to.